src/cls/IPM/General/History.cls

```diff
@@ -0,0 +1,524 @@
+}
+
+ClassMethod GlobalRecordsFetch(ByRef qHandle As %Binary, ByRef Row As %List, ByRef AtEnd As %Integer = 0) As %Status [ Internal, PlaceAfter = GlobalRecordsExecute ]
+{
+    Try {
+        While qHandle.Count() > 0 {
+            // Always get the first namespace. Remove it from list once corresponding result set is exhausted.
+            Set ns = ""
+            Set rs = qHandle.GetNext(.ns)
+            $$$ThrowOnError(..RecordsFetch(rs, .Row, .AtEnd))
+            If AtEnd {
+                // If rs is empty, there could still be more namespaces to process
+                Do qHandle.RemoveAt(ns, .success)
+                If 'success {
+                    $$$ThrowStatus($$$ERROR($$$GeneralError, "Failed to remove namespace "_ns_" from list"))
+                }
+            } Else {
+                Return $$$OK
+            }
+        }
+        // Only when the list is empty should we return AtEnd = 1 and Row = ""
+        Set Row = "", AtEnd = 1
+        Return $$$OK
+    } Catch ex {
+        Return ex.AsStatus()
+    }
+    Return $$$OK
+}
+
+ClassMethod GlobalRecordsClose(ByRef qHandle As %Binary) As %Status [ Internal, PlaceAfter = GlobalRecordsFetch ]
+{
+    // Should not happen, but just in case
+    If qHandle.Count() '= 0 {
+        $$$ThrowStatus($$$ERROR($$$GeneralError, "Namespace list not empty in GlobalRecordsFetch()!"))
+    }
+    Set qHandle = ""
+    Quit $$$OK
+}
+
+/// Get the history of all installations, uninstalls, and loads in all namespaces
+/// See GetHistory() for the `filter` argument
+ClassMethod GetHistoryGlobally(ByRef filter, ascend As %Boolean = 0, limit As %Integer = 0) As %Library.ArrayOfObjects
+{
+    Set originalNS = $Namespace
+    New $Namespace
+    Set $Namespace = "%SYS"
+    Set rs = ##class(%SQL.Statement).%ExecDirect(, "SELECT DISTINCT Nsp FROM %SYS.Namespace_List()")
+    $$$ThrowSQLIfError(rs.%SQLCODE, rs.%Message)
+    Set $Namespace = originalNS
+    Set array = ##class(%Library.ArrayOfObjects).%New()
+    While rs.%Next() {
+        Set ns = rs.%Get("Nsp")
+        Set historyRS = ..GetHistory(.filter, ascend, limit, ns)
+        If historyRS = $$$NULLOREF {
+            Continue
+        }
+        Do array.SetAt(historyRS, ns)
+    }
+    Return array
+}
+
+ClassMethod DisplayOneRecord(id As %Integer)
+{
+    Set rs = ##class(%SQL.Statement).%ExecDirect(, "SELECT TOP 1 * FROM %IPM_General.History WHERE ID = ?", id)
+    If rs.%Next() {
+        Do ..ShowColumns(rs, 1)
+    } Else {
+        $$$ThrowStatus($$$ERROR($$$GeneralError, "No record found with ID "_id_". Use `history find` command to show all entries. "))
+    }
+}
+
+/// If detail is 1, will display full details each property on a separate line
+/// If detail is "header", will display the column names using the same style as the data
+/// Otherwise, will display on a single line, with some details omitted
+ClassMethod ShowColumns(rs As %SQL.StatementResult, detail As %Boolean = 0) As %Status [ Internal, Private ]
```

Comment:
Rather than having detail be a fake boolean, perhaps have it be an enum? So valid options could be "header" and "full" perhaps?